invisible_cities/cities/detsim_simulate_electrons.py

```diff
@@ -0,0 +1,140 @@
+
+    Parrameters:
+        :wi: float
+            ionization yield
+        :fano_factor: float
+            fano-factor. related with the deviation in ionization electrons
+        :energies: np.ndarray
+            energy hits
+    Returns:
+        :nes: np.ndarray
+            the ionization electrons per hit
+    """
+    nes  = np.array(energies/wi, dtype = int)
+    pois = nes < 10
+    nes[ pois] = np.random.poisson(nes[pois])
+    nes[~pois] = np.round(np.random.normal(nes[~pois], np.sqrt(nes[~pois] * fano_factor)))
+    return nes
+
+
+def drift_electrons(lifetime       : float,
+                    drift_velocity : float,
+                    zs             : np.ndarray,
+                    electrons      : np.ndarray) -> np.array:
+    """ returns number of electrons due to lifetime loses from secondary electrons
+
+    Parameters:
+        :lifetime: float
+            electron lifetime
+        :drif_velocity: float
+            drif velocity at the active volume of the detector
+        :zs:
+            the z hits
+        :electrons:
+    Returns:
+        :nes: np.ndarray
+            the drifted electrons per initial hit
+    """
+    ts  = zs / drift_velocity
+    nes = electrons - np.random.poisson(electrons * (1. - np.exp(-ts/lifetime)))
+    nes[nes < 0] = 0
+    return nes
+
+
+def diffuse_electrons(transverse_diffusion   : float,
+                      longitudinal_diffusion : float,
+                      xs                     : np.ndarray,
+                      ys                     : np.ndarray,
+                      zs                     : np.ndarray,
+                      electrons              : np.ndarray)\
+                      -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    starting from hits with positions xs, ys, zs, and number of electrons,
+    apply diffusion and return diffused positions xs, ys, zs for each electron.
+
+    Paramters:
+        :transverse_diffusion: float
+        :longitudinal_diffusion: float
+        :xs, ys, zs: np.ndarray (1D of size: #hits)
+            postions of initial hits
+        :electrons:
+            electrons at the EL before drifting per hit
+    Returns:
+        :dxs, dys, dzs: np.ndarray (1D of size: #hits x #total electrons)
+            diffused positions
+    """
+    xs = np.repeat(xs, electrons.astype(int))
+    ys = np.repeat(ys, electrons.astype(int))
+    zs = np.repeat(zs, electrons.astype(int))
+
+    # substitute z<0 to z=0
+    sel = zs<0
+    zs[sel] = 0
+
+    sqrtz = zs ** 0.5
+    dxs  = np.random.normal(xs, sqrtz * transverse_diffusion)
+    dys  = np.random.normal(ys, sqrtz * transverse_diffusion)
+    dzs  = np.random.normal(zs, sqrtz * longitudinal_diffusion)
+
+    return (dxs, dys, dzs)
+
+
+def voxelize(voxel_size : list,
+             dx         : np.ndarray,
+             dy         : np.ndarray,
+             dz         : np.ndarray)\
+             ->Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```

Comment:
If you look at `pes_at_sipms` function in PR #734, for each hit at the EL we compute the distance between this hit and each of the sensors (actually not all the sensors, but the sensors in a framed region at TP).
So this matrix (`distances`) has shape (nhits, nsensors). Now, for each distance in the matrix, we need to go to the PSF to take the corresponding value. PSF is not a single number, since we have PSFs for several EL partitions, so the PSF values would be a matrix of shape (nhits, nsensors, npartitions). 
This matrix becames huge for high energy events, since number of hits is around 100K.
The voxelization greatly reduces the number of hits, and the the size of this matrix.

Reply:
Probably we dont want voxelization of ionization electrons at the EL anymore, since I think we can use the loop-based implementation for SIPM signal instead the matrix-based. I would left the function here until we decide.